backend/src/main/java/fk/prof/backend/model/association/impl/ZookeeperBasedBackendAssociationStore.java

```diff
@@ -0,0 +1,336 @@
+      for(Recorder.ProcessGroup processGroup: backendDetail.getAssociatedProcessGroups()) {
+        if (this.processGroupToBackendLookup.putIfAbsent(processGroup, backendDetail.getBackendIPAddress()) != null) {
+          throw new IllegalStateException(String.format("Backend mapping already exists for process group=%s", processGroup));
+        }
+      }
+    }
+  }
+
+  @Override
+  public Future<Recorder.ProcessGroups> reportBackendLoad(BackendDTO.LoadReportRequest payload) {
+    String backendIPAddress = payload.getIp();
+    Future<Recorder.ProcessGroups> result = Future.future();
+    vertx.executeBlocking(future -> {
+      try {
+        BackendDetail backendDetail = backendDetailLookup.computeIfAbsent(backendIPAddress, (key) -> {
+          try {
+            BackendDetail updatedBackendDetail = new BackendDetail(key, reportingFrequencyInSeconds, maxAllowedSkips);
+            String zNodePath = getZNodePathForBackend(key);
+            ZookeeperUtil.writeZNode(curatorClient, zNodePath, BackendDetail.serializeProcessGroups(updatedBackendDetail.getAssociatedProcessGroups()), true);
+            return updatedBackendDetail;
+          } catch (Exception ex) {
+            throw new RuntimeException(ex);
+          }
+        });
+
+        /**
+         * If backend was defunct earlier, acquire lock on available backend queue and add this backend there along with reporting load
+         * If backend was available, don't incur cost of acquiring lock since backend will already be in queue, which is the case mostly. Report load and get out
+         */
+        boolean wasDefunct = backendDetail.isDefunct();
+        if(wasDefunct) {
+          try {
+            boolean acquired = backendAssignmentLock.tryLock(2, TimeUnit.SECONDS);
+            if (acquired) {
+              backendDetail.reportLoad(payload.getLoad());
+              availableBackendsByPriority.offer(backendDetail);
+            } else {
+              logger.warn("Timeout while acquiring lock on backend queue for reporting backend=" + backendIPAddress);
+            }
+          } catch (InterruptedException ex) {
+            logger.warn("Interrupted while acquiring lock on backend queue for reporting backend=" + backendIPAddress, ex);
+          } finally {
+            backendAssignmentLock.unlock();
+          }
+        } else {
+          backendDetail.reportLoad(payload.getLoad());
+        }
+        future.complete(BackendDetail.buildProcessGroupsProto(backendDetail.getAssociatedProcessGroups()));
+      } catch (Exception ex) {
+        future.fail(ex);
+      }
+
+    }, result.completer());
+    return result;
+  }
+
+  @Override
+  public Future<String> getAssociatedBackend(Recorder.ProcessGroup processGroup) {
+    Future<String> result = Future.future();
+    String existingBackendIPAddress = processGroupToBackendLookup.get(processGroup);
+    vertx.executeBlocking(future -> {
+      if(existingBackendIPAddress != null
```

Comment:
No it does not, but since else requires to happen in async, wrapped the whole thing as one async operation. There should not be any major cost associated with this, async operation is scheduled on same event loop(thread)